File: PancakeMusicProjectsTS/src/model/guitar/Tuning.ts
import { Note } from "../musictheory/Note.js";
import { NoteAndOctave } from "../musictheory/NoteAndOctave.js";
import { EqualsTrait } from "../general/EqualsTrait.js";

export class Tuning implements EqualsTrait {

  public static STANDARD_7_STRING_GUITAR = new Tuning("STANDARD_7_STRING_GUITAR", [
    new NoteAndOctave(Note.B, 2),
    new NoteAndOctave(Note.E, 2),
    new NoteAndOctave(Note.A, 3),
    new NoteAndOctave(Note.D, 3),
    new NoteAndOctave(Note.G, 3),
    new NoteAndOctave(Note.B, 4),
    new NoteAndOctave(Note.E, 4)
  ]);

  public static DROP_A_7_STRING_GUITAR = new Tuning("DROP_A_7_STRING_GUITAR", [
    new NoteAndOctave(Note.A, 2),
    new NoteAndOctave(Note.E, 2),
    new NoteAndOctave(Note.A, 3),
    new NoteAndOctave(Note.D, 3),
    new NoteAndOctave(Note.G, 3),
    new NoteAndOctave(Note.B, 4),
    new NoteAndOctave(Note.E, 4)
  ]);

  public static STANDARD_6_STRING_GUITAR = new Tuning("STANDARD_6_STRING_GUITAR", [
    new NoteAndOctave(Note.E, 2),
    new NoteAndOctave(Note.A, 3),
    new NoteAndOctave(Note.D, 3),
    new NoteAndOctave(Note.G, 3),
    new NoteAndOctave(Note.B, 4),
    new NoteAndOctave(Note.E, 4)
  ]);

  public static DROP_D_6_STRING_GUITAR = new Tuning("DROP_D_6_STRING_GUITAR", [
    new NoteAndOctave(Note.D, 2),
    new NoteAndOctave(Note.A, 3),
    new NoteAndOctave(Note.D, 3),
    new NoteAndOctave(Note.G, 3),
    new NoteAndOctave(Note.B, 4),
    new NoteAndOctave(Note.E, 4)
  ]);

  public static DADGAD_6_STRING_GUITAR = new Tuning("DADGAD_6_STRING_GUITAR", [
    new NoteAndOctave(Note.D, 2),
    new NoteAndOctave(Note.A, 3),
    new NoteAndOctave(Note.D, 3),
    new NoteAndOctave(Note.G, 3),
    new NoteAndOctave(Note.A, 4),
    new NoteAndOctave(Note.D, 4)
  ]);

  public static STANDARD_4_STRING_BASS = new Tuning("STANDARD_4_STRING_BASS", [
    new NoteAndOctave(Note.E, 1),
    new NoteAndOctave(Note.A, 2),
    new NoteAndOctave(Note.D, 2),
    new NoteAndOctave(Note.G, 2)
  ]);

  public static STANDARD_5_STRING_BASS = new Tuning("STANDARD_5_STRING_BASS", [
    new NoteAndOctave(Note.B, 1),
    new NoteAndOctave(Note.E, 1),
    new NoteAndOctave(Note.A, 2),
    new NoteAndOctave(Note.D, 2),
    new NoteAndOctave(Note.G, 2)
  ]);

  public static STANDARD_6_STRING_BASS = new Tuning("STANDARD_6_STRING_BASS", [
    new NoteAndOctave(Note.B, 1),
    new NoteAndOctave(Note.E, 1),
    new NoteAndOctave(Note.A, 2),
    new NoteAndOctave(Note.D, 2),
    new NoteAndOctave(Note.G, 2),
    new NoteAndOctave(Note.C, 3),
  ]);

  public static STANDARD_4_COURSES_MANDOLIN = new Tuning("STANDARD_4_COURSES_MANDOLIN", [
    new NoteAndOctave(Note.G, 3),
    new NoteAndOctave(Note.D, 4),
    new NoteAndOctave(Note.A, 4),
    new NoteAndOctave(Note.E, 5)
  ]);

  public static STANDARD_4_STRING_UKULELE = new Tuning("STANDARD_4_STRING_UKULELE", [
    new NoteAndOctave(Note.G, 4),
    new NoteAndOctave(Note.C, 4),
    new NoteAndOctave(Note.E, 4),
    new NoteAndOctave(Note.A, 4)
  ]);

  private static addNameAndValue(map: Map<string, Tuning>, tuning: Tuning): void {
    map.set(tuning.name, tuning);
  }

  private static initAvailableTunings() : Map<string, Tuning> {
    const map: Map<string, Tuning> = new Map();

    Tuning.addNameAndValue(map, Tuning.STANDARD_6_STRING_GUITAR);
    Tuning.addNameAndValue(map, Tuning.STANDARD_7_STRING_GUITAR);
    Tuning.addNameAndValue(map, Tuning.DROP_D_6_STRING_GUITAR);
    Tuning.addNameAndValue(map, Tuning.DROP_A_7_STRING_GUITAR);
    Tuning.addNameAndValue(map, Tuning.DADGAD_6_STRING_GUITAR);
    Tuning.addNameAndValue(map, Tuning.STANDARD_4_STRING_BASS);
    Tuning.addNameAndValue(map, Tuning.STANDARD_5_STRING_BASS);
    Tuning.addNameAndValue(map, Tuning.STANDARD_6_STRING_BASS);
    Tuning.addNameAndValue(map, Tuning.STANDARD_4_COURSES_MANDOLIN);
    Tuning.addNameAndValue(map, Tuning.STANDARD_4_STRING_UKULELE);

    return map;
  }

  public static availableTunings : Map<string, Tuning> = Tuning.initAvailableTunings();


  public static getTuningByName(name: string): Tuning | null {
    for (const [tuningName, tuning] of Tuning.availableTunings) {
      if (tuningName === name) {
        return tuning;
      }
    }
    return null;
  }

  readonly stringNotes: Map<number, NoteAndOctave>;
  readonly name: string;

  constructor(name: string, stringTunings: NoteAndOctave[]) {
    this.name = name;
    let map: Map<number, NoteAndOctave> = new Map();
    let i = 0;
    for (const stringTuning of stringTunings) {
      map.set(i, stringTuning);
      i++;
    }

    this.stringNotes = map;
  }

  public getStringNotes(): Map<number, NoteAndOctave> {
    return this.stringNotes;
  }

  public getNumberOfStrings(): number {
    return this.stringNotes.size;
  }

  equals(other: any): boolean {

    if (!(other instanceof Tuning)) {
      return false;
    }

    if (this.stringNotes.size !== other.stringNotes.size) {
      return false;
    }

    for (const [guitarString, note] of this.stringNotes) {
      const otherNote = other.stringNotes.get(guitarString);

      if (!otherNote) {
        return false;
      }

      if (note !== otherNote && !note.equals(otherNote)) {
        return false;
      }
    }

    return true;
  }

  toString(): string {
    const stringArray: string[] = [];
    for (const [gs, stringNote] of this.stringNotes) {
      stringArray.push("" + gs + "=" + stringNote.toString());
    }
    return stringArray.join(' ');
  }
}

File: ChordVoicingsCreatorJava/src/main/java/com/campoy/chord/voicing/creator/model/guitar/Tuning.java
package com.campoy.chord.voicing.creator.model.guitar;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

import com.campoy.chord.voicing.creator.model.musictheory.Note;
import com.campoy.chord.voicing.creator.model.musictheory.NoteAndOctave;

import java.util.StringJoiner;

public class Tuning {

    public static final Tuning STANDARD_7_STRING_GUITAR = new Tuning();
    public static final Tuning DROP_A_7_STRING_GUITAR = new Tuning();
    public static final Tuning STANDARD_6_STRING_GUITAR = new Tuning();
    public static final Tuning DROP_D_6_STRING_GUITAR = new Tuning();

    public static final Tuning STANDARD_6_STRING_BASS = new Tuning();
    public static final Tuning STANDARD_5_STRING_BASS = new Tuning();
    public static final Tuning STANDARD_4_STRING_BASS = new Tuning();
    
    public static final Tuning[] presetTunings = {
            STANDARD_6_STRING_GUITAR,
            DROP_D_6_STRING_GUITAR,
            STANDARD_7_STRING_GUITAR,
            DROP_A_7_STRING_GUITAR,
            STANDARD_4_STRING_BASS,
            STANDARD_5_STRING_BASS,
            STANDARD_6_STRING_BASS };
    
    static {

        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.B, 2));
        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.E, 2));
        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.A, 3));
        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.D, 3));
        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(4),
                new NoteAndOctave(Note.G, 3));
        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(5),
                new NoteAndOctave(Note.B, 4));
        STANDARD_7_STRING_GUITAR.stringNotes.put(new GuitarString(6),
                new NoteAndOctave(Note.E, 4));
        
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.A, 2));
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.E, 2));
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.A, 3));
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.D, 3));
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(4),
                new NoteAndOctave(Note.G, 3));
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(5),
                new NoteAndOctave(Note.B, 4));
        DROP_A_7_STRING_GUITAR.stringNotes.put(new GuitarString(6),
                new NoteAndOctave(Note.E, 4));
        
        STANDARD_6_STRING_GUITAR.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.E, 2));
        STANDARD_6_STRING_GUITAR.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.A, 3));
        STANDARD_6_STRING_GUITAR.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.D, 3));
        STANDARD_6_STRING_GUITAR.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.G, 3));
        STANDARD_6_STRING_GUITAR.stringNotes.put(new GuitarString(4),
                new NoteAndOctave(Note.B, 4));
        STANDARD_6_STRING_GUITAR.stringNotes.put(new GuitarString(5),
                new NoteAndOctave(Note.E, 4));
        
        DROP_D_6_STRING_GUITAR.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.D, 2));
        DROP_D_6_STRING_GUITAR.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.A, 3));
        DROP_D_6_STRING_GUITAR.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.D, 3));
        DROP_D_6_STRING_GUITAR.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.G, 3));
        DROP_D_6_STRING_GUITAR.stringNotes.put(new GuitarString(4),
                new NoteAndOctave(Note.B, 4));
        DROP_D_6_STRING_GUITAR.stringNotes.put(new GuitarString(5),
                new NoteAndOctave(Note.E, 4));
        
        STANDARD_4_STRING_BASS.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.E, 1));
        STANDARD_4_STRING_BASS.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.A, 2));
        STANDARD_4_STRING_BASS.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.D, 2));
        STANDARD_4_STRING_BASS.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.G, 2));
        
        STANDARD_5_STRING_BASS.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.B, 1));
        STANDARD_5_STRING_BASS.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.E, 1));
        STANDARD_5_STRING_BASS.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.A, 2));
        STANDARD_5_STRING_BASS.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.D, 2));
        STANDARD_5_STRING_BASS.stringNotes.put(new GuitarString(4),
                new NoteAndOctave(Note.G, 2));
        
        STANDARD_6_STRING_BASS.stringNotes.put(new GuitarString(0),
                new NoteAndOctave(Note.B, 1));
        STANDARD_6_STRING_BASS.stringNotes.put(new GuitarString(1),
                new NoteAndOctave(Note.E, 1));
        STANDARD_6_STRING_BASS.stringNotes.put(new GuitarString(2),
                new NoteAndOctave(Note.A, 2));
        STANDARD_6_STRING_BASS.stringNotes.put(new GuitarString(3),
                new NoteAndOctave(Note.D, 2));
        STANDARD_6_STRING_BASS.stringNotes.put(new GuitarString(4),
                new NoteAndOctave(Note.G, 2));
        STANDARD_6_STRING_BASS.stringNotes.put(new GuitarString(5),
                new NoteAndOctave(Note.C, 3));

    }
    
    private final Map<GuitarString, NoteAndOctave> stringNotes = new HashMap<>();
    
    private Tuning() {
    }
    
    public Tuning(Map<GuitarString, NoteAndOctave> stringNotes) {
        this.stringNotes.putAll(stringNotes);
    }
    
    public Map<GuitarString, NoteAndOctave> getStringNotes() {
        return Collections.unmodifiableMap(stringNotes);        
    }
    
    public int getNumberOfStrings(){
        return stringNotes.size();
    }
    
    @Override
    public boolean equals(Object obj) {
        
        if( !( obj instanceof Tuning) ) {
            return false;
        }
        
        Tuning objTuning = (Tuning)obj;
        
        if(this.stringNotes.size() != objTuning.stringNotes.size()) {
            return false;
        }
                
        for(Entry<GuitarString, NoteAndOctave> entry : this.stringNotes.entrySet()) {
            
            GuitarString guitarString = entry.getKey();
            NoteAndOctave stringNote = entry.getValue();
            NoteAndOctave otherStringNote = objTuning.stringNotes.get(guitarString);
            
            if(stringNote != otherStringNote 
                    && !stringNote.equals(otherStringNote)) {
                return false;
            }
        }
        
        return true;
    }
    
    @Override
    public String toString() {
        
        StringJoiner sj = new StringJoiner(" "); 
        for(int i = 0; i < stringNotes.size(); i++) {
            NoteAndOctave stringNote = stringNotes.get(new GuitarString(i));
            sj.add(stringNote.toString());
        }
        return sj.toString();
    }
}
